network/src/worker.rs

```diff
@@ -69,21 +68,25 @@ impl WorkerNetwork {
         &mut self,
         address: Multiaddr,
         message: BincodeEncodedPayload,
-    ) -> CancelHandler<()> {
+    ) -> CancelHandler<MessageResult> {
         let client = self.client(address);
+
+        let message_send = move || {
+            let mut client = client.clone();
+            let message = message.clone();
+
+            async move {
+                client.send_message(message).await.map_err(|e| {
+                    // this returns a backoff::Error::Transient
+                    // so that if tonic::Status is returned, we retry
+                    Into::<backoff::Error<anyhow::Error>>::into(anyhow::Error::from(e))
+                })
+            }
+        };
+
         let handle = self
             .executor
-            .spawn(
-                self.retry_config
-                    .retry(move || {
-                        let mut client = client.clone();
-                        let message = message.clone();
-                        async move { client.send_message(message).await.map_err(Into::into) }
-                    })
-                    .map(|response| {
-                        response.expect("we retry forever so this shouldn't fail");
-                    }),
-            )
+            .spawn_with_retries(self.retry_config, message_send)
```

Comment:
This does create more tasks right?

Reply:
Overall, yes. But it does not create more concurrent network messages.